### ISSUE TITLE
Verbalize all reasoning traces in latent space representations in each intermediate latent reasoning step into human language?

### ISSUE THREAD

**Issue by jbdatascience** [seal-rg/recurrent-pretraining#6]
A potential drawback of this recurrent reasoning process in latent space is the opaque reasoning process. But my view on this drawbck would be, that it is possible to apply the Coda block (that verbalizes the latent space results to normal human interpretable language ) to each intermediate reasoning step in latent space! Can you give a comprehensive and detailed assesment of my view.

I mean applying the Coda block at each intermediate reasoning step in latent space, but in the mean time letting the reasoning steps in latent space go on unchanged (not being influenced by the Coda block data). Can you give your assessment on that?

Of course this would mean much more compute that would be needed, but one could use this only for research purposes. 

Also maybe not all reasoning traces in latent space representations are verbalizable into normal human language?

**Comment by JonasGeiping** [comment]
Hi, this sounds sensible. It would help provide feedback on the model's "current prediction" at any step in the trajectory. 

I've looked these trajectories for some of these on GSM8k questions (mostly because the coda block+head can also be used to decide early exits). On math questions, I've observed that for crucial tokens, often nothing seems to happen, the model predicts the  wrong answer for a number of steps, and then flips to a correct answer. So, if can only observe only the current answer of the model, and not always its current thought process.

**Comment by JonasGeiping** [comment]
No, I haven't observed that for the prompts I looked at (but I've only started to really explore this model in depth)

**Comment by jbdatascience** [comment]
Very interesting developments! 
Have I understand it correctly that by means of recurrent pre-training the model will generate reasoning traces in latent space that converge to a fixed point in latent space? And that this fixed point represents the optimal response to the input query?

**Comment by JonasGeiping** [comment]
Not necessarily a fixed point, we also observe other terminal behaviors, such as orbits (see Section 7, and the Appendix)

**Comment by jbdatascience** [comment]
> Not necessarily a fixed point, we also observe other terminal behaviors, such as orbits (see Section 7, and the Appendix)

Are there any (theoretical) guarentees that the model will converge to the best possible results?

**Comment by JonasGeiping** [comment]
This has been explored in some of the previous work, but in our work, we did not want to impose a particular convergent model, but learn terminal behaviors directly from data. The loss incentives a final model where it is likely that decoding a late latent state results in a good next-token prediction.

**Comment by jbdatascience** [comment]
> This has been explored in some of the previous work, but in our work, we did not want to impose a particular convergent model, but learn terminal behaviors directly from data. The loss incentives a final model where it is likely that decoding a late latent state results in a good next-token prediction, but optimality is not guaranteed.

I am very curious where this development will lead to eventually! Thank you for your very and quick responses. I will follow this closely!
